jdaviz/utils.py

```diff
@@ -385,3 +391,132 @@ def total_masked_first_data(self):
+    cache: None, bool, or ``"update"``, optional
+        Cache file after download. If ``possible_uri`` is a
+        URL, ``cache`` may be a boolean or ``"update"``, see documentation for
+        `~astropy.utils.data.download_file` for details. If cache is None,
+        the file is cached and a warning is raised suggesting to set ``cache``
+        explicitly in the future.
+    local_path : str, optional
+        Save the downloaded file to this path. Default is to
+        save the file with its remote filename in the current
+        working directory. This is only used if data is requested
+        from `astroquery.mast`.
+    timeout : float, optional
+        If downloading from a remote URI, set the timeout limit for
+        remote requests in seconds (passed to
+        `~astropy.utils.data.download_file` or
+        `~astroquery.mast.Conf.timeout`).
+
+    Returns
+    -------
+    possible_uri : str or other
+        If ``possible_uri`` cannot be retrieved as a URI, returns the input argument
+        unchanged. If ``possible_uri`` can be retrieved as a URI, returns the
+        local path to the downloaded file.
+    """
+
+    if os.environ.get("JDAVIZ_START_DIR", ""):
+        # avoiding creating local paths in a tmp dir when in standalone:
+        local_path = os.environ["JDAVIZ_START_DIR"] / local_path
+
+    if not isinstance(possible_uri, str):
+        # only try to parse strings:
+        return possible_uri
+
+    if os.path.exists(possible_uri):
+        # don't try to parse file paths:
+        return possible_uri
+
+    parsed_uri = urlparse(possible_uri)
+
+    cache_none_msg = (
+        "You may be querying for a remote file "
+        f"at '{possible_uri}', but the `cache` argument was not "
+        f"in the call to `load_data`. Unless you set `cache` "
+        f"explicitly, remote files will be cached locally and "
+        f"this warning will be raised."
+    )
+
+    local_path_msg = (
+        f"You requested to cache data to the `local_path`='{local_path}'. This "
+        f"keyword argument is supported for downloads of MAST URIs via astroquery, "
+        f"but since the remote file at '{possible_uri}' will be downloaded "
+        f"using `astropy.utils.data.download_file`, the file will be "
+        f"stored in the astropy download cache instead."
+    )
+
+    cache_warning = False
+    if cache is None:
+        cache = True
+        cache_warning = True
+
+    if parsed_uri.scheme.lower() == 'mast':
+        if cache_warning:
+            warnings.warn(cache_none_msg, UserWarning)
+
+        if local_path is not None and os.path.isdir(local_path):
+            # if you give a directory, save the file there with default name:
+            local_path = os.path.join(local_path, parsed_uri.path.split(os.path.sep)[-1])
+
+        if timeout is None:
+            timeout = Conf.timeout.defaultvalue
+
+        with Conf.timeout.set_temp(timeout):
```

Comment:
```suggestion
        with conf.timeout.set_temp(timeout):
```

Reply:
Thanks, just caught these too in 77df5ff.

Reply:
If we want to be more pedantic, we could `from astroquery.mast import conf as mast_conf` but since there isn't any name clash now, doesn't matter.